### PR TITLE
Added two new layouts for the cospad

### PR DESCRIPTION
**Pull request by johannbl**
## Description

Two new numpad layouts for the cospad focusing on replacing the 2u "0" key with two 1u keys: "00" and "0".
A new keymap using numpad2 layout is provided. It includes the "00" key created using a SEND_STRING() macro.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#include QMK_KEYBOARD_H
+
+/* THIS FILE WAS GENERATED!
+ *
+ * This file was generated by qmk-compile-json. You may or may not want to
+ * edit it directly.
+ */
+
```
Comment (tzarc) [inline keyboards/cospad/keymaps/johannbl/keymap.c]: Let's remove these now that it's been edited.

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        ),
+        
+        [1] = LAYOUT_numpad2_6x4(
+        	RGB_TOG, RGB_RMOD, RGB_MOD, KC_TRNS, 
+        	KC_NO, RGB_SPD, RGB_SPI, KC_NO, 
+        	RGB_M_G, RGB_M_SW, RGB_M_K, RGB_M_SN, 
+        	RGB_M_X, RGB_M_R, RGB_M_P, RGB_M_B, 
+        	RGB_HUI, RGB_SAI, RGB_VAI, 
+        	RGB_HUD, RGB_SAD, RGB_VAD, RESET
+        ),
```
Comment (tzarc) [inline keyboards/cospad/keymaps/johannbl/keymap.c]: How is this layer activated?

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
+#include QMK_KEYBOARD_H
+
+/* THIS FILE WAS GENERATED!
+ *
+ * This file was generated by qmk-compile-json. You may or may not want to
+ * edit it directly.
+ */
```
Comment (drashna) [inline keyboards/cospad/keymaps/johannbl/keymap.c]:
```suggestion
```

File: keyboards/cospad/keymaps/johannbl/keymap.c (hunk 2 above)
Comment (johannbl) [reply keyboards/cospad/keymaps/johannbl/keymap.c]: Good point. I can't say the keymap file is very clean. This is the one I use at the moment and I don't care about backlighting. I could instead propose 2 more standard numpad keymap. One for each layout.

**Comment by johannbl** [comment]
I added two new keymap for the new layouts. They are in their own folder. I left my own default keymap there but I would not merge it with the main project.

You can see this here: https://github.com/johannbl/qmk_firmware/tree/master/keyboards/cospad/keymaps/johannbl

I'm not sure how I can pull request specific files rather than the whole thing.

File: keyboards/cospad/cospad.h [hunk 4]
```diff
@@ -125,6 +125,70 @@
...
+ * | 20 | 21 | 22 | 23 |
+ * |----|----|----|----|
+ * | 30 | 31 | 32 | 33 |
+ * |----|----|----|----|
+ * | 40 | 41 | 42 |    |
+ * |----|----|----| 43 |
+ * | 50 | 51 | 52 |    |
+ * `-------------------'
+ */
+#define LAYOUT_numpad2_6x4( \
```
Comment (noroadsleft) [inline keyboards/cospad/cospad.h]: It's wordy, but more descriptive and more in line with our current conventions to name this as:

```suggestion
/* 6x4 numpad with split Plus and 0 keys
 * ,-------------------.
 * | 00 | 01 | 02 | 03 |
 * |----|----|----|----|
 * | 10 | 11 | 12 | 13 |
 * |----|----|----|----|
 * | 20 | 21 | 22 | 23 |
 * |----|----|----|----|
 * | 30 | 31 | 32 | 33 |
 * |----|----|----|----|
 * | 40 | 41 | 42 |    |
 * |----|----|----| 43 |
 * | 50 | 51 | 52 |    |
 * `-------------------'
 */
#define LAYOUT_numpad_6x4_split_plus_zero( \
```

File: keyboards/cospad/cospad.h [hunk 5]
```diff
@@ -125,6 +125,70 @@
...
+ * | 20 | 21 | 22 |    |
+ * |----|----|----| 23 |
+ * | 30 | 31 | 32 |    |
+ * |----|----|----|----|
+ * | 40 | 41 | 42 |    |
+ * |----|----|----| 43 |
+ * | 50 | 51 | 52 |    |
+ * `-------------------'
+ */
+#define LAYOUT_numpad3_6x4( \
```
Comment (noroadsleft) [inline keyboards/cospad/cospad.h]:
```suggestion
/* 6x4 numpad with split 0 key
 * ,-------------------.
 * | 00 | 01 | 02 | 03 |
 * |----|----|----|----|
 * | 10 | 11 | 12 | 13 |
 * |----|----|----|----|
 * | 20 | 21 | 22 |    |
 * |----|----|----| 23 |
 * | 30 | 31 | 32 |    |
 * |----|----|----|----|
 * | 40 | 41 | 42 |    |
 * |----|----|----| 43 |
 * | 50 | 51 | 52 |    |
 * `-------------------'
 */
#define LAYOUT_numpad_6x4_split_zero( \
```

File: keyboards/cospad/info.json [hunk 6]
```diff
@@ -102,6 +102,71 @@
                 {"x":2, "y":5},
                 {"x":3, "y":5}
             ]
-        }
+        },
+        "LAYOUT_numpad2_6x4": {
```
Comment (noroadsleft) [inline keyboards/cospad/info.json]:
```suggestion
        "LAYOUT_numpad_6x4_split_plus_zero": {
```

File: keyboards/cospad/info.json [hunk 7]
```diff
@@ -102,6 +102,71 @@
...
+                {"x":1, "y":4},
+                {"x":2, "y":4},
+
+                {"x":0, "y":5},
+                {"x":1, "y":5},
+                {"x":2, "y":5},
+                {"x":3, "y":4, "h":2}
+            ]
+        },
+        "LAYOUT_numpad3_6x4": {
```
Comment (noroadsleft) [inline keyboards/cospad/info.json]:
```suggestion
        "LAYOUT_numpad_6x4_split_zero": {
```

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ 		* | 7  | 8  | 9  | -  |
+ 		* |----|----|----|----|
+ 		* | 4  | 5  | 6  | +  |
+ 		* |----|----|----|----|
+ 		* | 1  | 2  | 3  |    |
+ 		* |----|----|----| En |
+ 		* | 00 | 0  | .  |    |
+ 		* `-------------------'
+ 		*/
+        [0] = LAYOUT_numpad2_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/keymap.c]:
```suggestion
        [0] = LAYOUT_numpad_6x4_split_plus_zero(
```

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        	KC_MUTE, KC_VOLD, KC_VOLU, TT(2), 
+        	KC_TAB, KC_PSLS, KC_PAST, KC_BSPC, 
+        	KC_P7, KC_P8, KC_P9, KC_PMNS, 
+        	KC_P4, KC_P5, KC_P6, KC_PPLS, 
+        	KC_P1, KC_P2, KC_P3,
+        	DBL_ZRO, KC_P0, KC_PDOT, KC_PENT
+        ),
+        
+        [1] = LAYOUT_numpad2_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/keymap.c]:
```suggestion
        [1] = LAYOUT_numpad_6x4_split_plus_zero(
```

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 10]
```diff
@@ -0,0 +1,70 @@
+#include QMK_KEYBOARD_H
+
+/* 
+ *  This keymap is for the numpad2 layout focusing on the "00" key.
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/keymap.c]:
```suggestion
 *  This keymap is for the split Plus and split 0 layout focusing on the "00" key.
```

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        [1] = LAYOUT_numpad2_6x4(
+        	RGB_TOG, RGB_RMOD, RGB_MOD, KC_TRNS, 
+        	KC_NO, RGB_SPD, RGB_SPI, KC_NO, 
+        	RGB_M_G, RGB_M_SW, RGB_M_K, RGB_M_SN, 
+        	RGB_M_X, RGB_M_R, RGB_M_P, RGB_M_B, 
+        	RGB_HUI, RGB_SAI, RGB_VAI, 
+        	RGB_HUD, RGB_SAD, RGB_VAD, RESET
+        ),
+        
+        [2] = LAYOUT_numpad2_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/keymap.c]:
```suggestion
        [2] = LAYOUT_numpad_6x4_split_plus_zero(
```

File: keyboards/cospad/keymaps/johannbl/numpad2/keymap.c [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+     * | 7  | 8  | 9  | -  |
+     * |----|----|----|----|
+     * | 4  | 5  | 6  | +  |
+     * |----|----|----|----|
+     * | 1  | 2  | 3  |    |
+     * |----|----|----| En |
+     * | 0  | 00 | .  |    |
+     * `-------------------'
+     */
+    [_BL] = LAYOUT_numpad2_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad2/keymap.c]:
```suggestion
    [_BL] = LAYOUT_numpad_6x4_split_plus_zero(
```

File: keyboards/cospad/keymaps/johannbl/numpad2/keymap.c [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+     * |HUD |HUI |BON |    |
+     * |----|----|----|----|
+     * |SAD |SAI |BOFF|    |
+     * |----|----|----|----|
+     * |VAD |VAS |BSTP|    |
+     * |----|----|----|    |
+     * |    |    |RST |    |
+     * `-------------------'
+    */
+    [_FL] = LAYOUT_numpad2_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad2/keymap.c]:
```suggestion
    [_FL] = LAYOUT_numpad_6x4_split_plus_zero(
```

File: keyboards/cospad/keymaps/johannbl/numpad2/keymap.c [hunk 14]
```diff
@@ -0,0 +1,81 @@
+#include QMK_KEYBOARD_H
+
+/* 
+ *  This keymap is based on the default keymap for the cospad.
+ *  It was adapted for the numpad2 layout focusing on the "00" key.
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad2/keymap.c]:
```suggestion
 *  It was adapted for the split Plus and split 0 layout focusing on the "00" key.
```

File: keyboards/cospad/keymaps/johannbl/numpad3/keymap.c [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+     * | 7  | 8  | 9  |    |
+     * |----|----|----| +  |
+     * | 4  | 5  | 6  |    |
+     * |----|----|----|----|
+     * | 1  | 2  | 3  |    |
+     * |----|----|----| En |
+     * | 0  | 00 | .  |    |
+     * `-------------------'
+     */
+    [_BL] = LAYOUT_numpad3_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad3/keymap.c]:
```suggestion
    [_BL] = LAYOUT_numpad_6x4_split_zero(
```

File: keyboards/cospad/keymaps/johannbl/numpad3/keymap.c [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+     * |HUD |HUI |BON |    |
+     * |----|----|----|    |
+     * |SAD |SAI |BOFF|    |
+     * |----|----|----|----|
+     * |VAD |VAS |BSTP|    |
+     * |----|----|----|    |
+     * |    |    |RST |    |
+     * `-------------------'
+    */
+    [_FL] = LAYOUT_numpad3_6x4(
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad3/keymap.c]:
```suggestion
    [_FL] = LAYOUT_numpad_6x4_split_zero(
```

File: keyboards/cospad/keymaps/johannbl/numpad3/keymap.c [hunk 17]
```diff
@@ -0,0 +1,81 @@
+#include QMK_KEYBOARD_H
+
+/* 
+ *  This keymap is based on the default keymap for the cospad.
+ *  It was adapted for the numpad3 layout focusing on the "00" key.
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/numpad3/keymap.c]:
```suggestion
 *  It was adapted for the split 0 layout focusing on the "00" key.
```

**Comment by johannbl** [comment]
Ah yes of course. This is much more explicit. I have a question about the keymap names (folder names) They are still named numpad2 and numpad3. What do you think would be best?

File: keyboards/cospad/info.json [hunk 18]
```diff
@@ -102,6 +102,71 @@
...
+                {"x":2, "y":4},
+
+                {"x":0, "y":5},
+                {"x":1, "y":5},
+                {"x":2, "y":5},
+                {"x":3, "y":4, "h":2}
+            ]
+        },
+        "LAYOUT_numpad_6x4_split_zero": {
+            "key_count": 23,
```
Comment (noroadsleft) [inline keyboards/cospad/info.json]:
```suggestion
            "key_count": 22,
```

**Comment by johannbl** [comment]
> Still some outstanding issues from my last review, plus one thing I forgot to mark.

Yes. the keymap.c file in the root folder has an unused layer and I understand that it's not a good idea to add this keymap to the main project. This is my first pull request so I'm not sure exactly if it's possible to only include some files but I wouldn't add this one at all if it was possible. What is the easiest way to do this? Should I just remove it completely?

I think the numpad2 and numpad3 keymaps should be enough to get started with the new layouts.

File: keyboards/cospad/keymaps/johannbl/keymap.c [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        ),
+        
+        [2] = LAYOUT_numpad_6x4_split_plus_zero(
+        	KC_MPLY, KC_MPRV, KC_MNXT, TO(0), 
+        	KC_F13, KC_F14, KC_F15, KC_F16, 
+        	KC_F17, KC_F18, KC_F19, KC_F20, 
+        	KC_F21, KC_F22, KC_F23, KC_F24, 
+        	KC_NO, KC_NO, KC_NO,
+        	KC_WAKE, KC_CALC, KC_NO, RESET
+        )
```
Comment (noroadsleft) [inline keyboards/cospad/keymaps/johannbl/keymap.c]: This removes Layer 1, which you said you don't care about anyway. For your use, it won't change how the keymap behaves at all.

```suggestion
        [0] = LAYOUT_numpad_6x4_split_plus_zero(
        	KC_MUTE, KC_VOLD, KC_VOLU, TT(1), 
        	KC_TAB, KC_PSLS, KC_PAST, KC_BSPC, 
        	KC_P7, KC_P8, KC_P9, KC_PMNS, 
        	KC_P4, KC_P5, KC_P6, KC_PPLS, 
        	KC_P1, KC_P2, KC_P3,
        	DBL_ZRO, KC_P0, KC_PDOT, KC_PENT
        ),
        
        [1] = LAYOUT_numpad_6x4_split_plus_zero(
        	KC_MPLY, KC_MPRV, KC_MNXT, TO(0), 
        	KC_F13, KC_F14, KC_F15, KC_F16, 
        	KC_F17, KC_F18, KC_F19, KC_F20, 
        	KC_F21, KC_F22, KC_F23, KC_F24, 
        	KC_NO, KC_NO, KC_NO,
        	KC_WAKE, KC_CALC, KC_NO, RESET
        )
```

**Comment by noroadsleft** [comment]
> Yes. the keymap.c file in the root folder has an unused layer and I understand that it's not a good idea to add this keymap to the main project. This is my first pull request so I'm not sure exactly if it's possible to only include some files but I wouldn't add this one at all if it was possible. What is the easiest way to do this? Should I just remove it completely?

We do accept user keymaps to the repository, if you don't want your keymap merged that's completely up to you.

If you want to remove your keymap from your pull request, you can run:

```shell
git rm --cached -- keyboards/cospad/keymaps/johannbl/keymap.c
```

from the command line. This will tell Git to stop tracking the file, but to leave it on your computer.

> I think the numpad2 and numpad3 keymaps should be enough to get started with the new layouts.

For this, we like to have the keymap names tied to the layout macro names. I think these should be named:

`keyboards/cospad/keymaps/split_plus_and_zero/keymap.c` (numpad2)
`keyboards/cospad/keymaps/split_zero/keymap.c` (numpad3)

You can do this in the command line with:

```shell
git mv -- keyboards/cospad/keymaps/johannbl/numpad2/keymap.c keyboards/cospad/keymaps/split_plus_and_zero/keymap.c
git mv -- keyboards/cospad/keymaps/johannbl/numpad3/keymap.c keyboards/cospad/keymaps/split_zero/keymap.c
```

If you're using GitHub's web interface, you can [follow these instructions](https://help.github.com/en/github/managing-files-in-a-repository/moving-a-file-to-a-new-location).

**Comment by johannbl** [comment]
Alright! Thank you so much. I got rid of the other keymap because I don't think it brings anything relevant.

File: keyboards/cospad/keymaps/split_plus_and_zero/keymap.c [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    /* Keymap _BL: (Base Layer) Default Layer
+     * ,-------------------.
+     * |Esc |TAB | FN | BS |
+     * |----|----|----|----|
+     * | NL | /  | *  | ~  |
+     * |----|----|----|----|
+     * | 7  | 8  | 9  | -  |
```
Comment (tzarc) [inline keyboards/cospad/keymaps/split_plus_and_zero/keymap.c]: Tilde and minus are the wrong way around, with respect to the layout:
```suggestion
     * | NL | /  | *  | -  |
     * |----|----|----|----|
     * | 7  | 8  | 9  | ~  |
```

**Pull request merged**